packages/network-controller/src/create-auto-managed-network-client.ts

```diff
@@ -0,0 +1,172 @@
+import { createNetworkClient, NetworkClient } from './create-network-client';
+import { BlockTracker, NetworkClientConfiguration, Provider } from './types';
+
+/**
+ * Represents a proxy object which wraps a target object. As a proxy, it allows
+ * for accessing and setting all of the properties that the target object
+ * supports, but also supports an extra propertyName (`__target__`) to access
+ * the target itself.
+ *
+ * @template Type - The type of the target object. It is assumed that this type
+ * will be constant even when the target is swapped.
+ */
+export type ProxyWithAccessibleTarget<TargetType> = TargetType & {
+  __target__: TargetType;
+};
+
+/**
+ * An object that provides the same interface as a network client but where the
+ * network client is not initialized until either the provider or block tracker
+ * is first accessed.
+ */
+export type AutoManagedNetworkClient<
+  Configuration extends NetworkClientConfiguration,
+> = {
+  configuration: Configuration;
+  provider: ProxyWithAccessibleTarget<Provider>;
+  blockTracker: ProxyWithAccessibleTarget<BlockTracker>;
+  destroy: () => void;
+};
+
+/**
+ * By default, the provider and block provider proxies will point to nothing.
+ * This is impossible when using the Proxy API, as the target object has to be
+ * something, so this object represents that "something".
+ */
+const UNINITIALIZED_TARGET = { __UNINITIALIZED__: true };
+
+/**
+ * The name of the method on both the provider and block tracker proxy which can
+ * be used to get the underlying provider or block tracker from the network
+ * client, when it is initialized.
+ */
+const REFLECTIVE_PROPERTY_NAME = '__target__';
+
+/**
+ * This function creates two proxies, one that wraps a provider and another that
+ * wraps a block tracker. These proxies are unique in that both will be "empty"
+ * at first; that is, neither will point to a functional provider or block
+ * tracker. They are set up in this way so as not to create unexpected network
+ * requests if the consumer of the proxy never accesses the network. Instead, as
```

Comment:
In theory, we could create the clients right away and we'd still have no extra/unexpected network requests. We don't get those until the client starts polling or until a request is made.

I was thinking of this proxy as being for the purpose of lazily performing the network initialization (e.g. checking network ID / latest block), which I see isn't being done yet.

Though... less stuff floating around in memory is good I suppose!

Reply:
Ah... I didn't make that connection. I'd have to think about how to make that work.

Reply:
Coming back to this again, it is true that merely making the network client construction lazy does not do what I advertise in the JSDocs. I was thinking that I would take it out and save it for another PR entirely, however, in attempting this, I realized that this change does have two benefits:

1. It saves me from having to update a ton of tests. There are a lot of tests that involve more than one network, and when they run they are creating fake providers for only the networks that they care about — they don't mock `createNetworkClient` for all of the possible built-in networks. It would take me a lot of time to change them, which unfortunately is not something I have at the moment. That said...
2. I think the design of the "auto-managed network client" here is okay. Perhaps it may need to be tweaked slightly once we make the "lookup network" code lazy too, but, the "special sauce" is a custom proxy, and I believe that will have to stay anyway.

So I will update the JSDocs to be more honest about what the proxy does at the moment.

Reply:
JSDocs updated here: [f2e5549](https://github.com/MetaMask/core/pull/1439/commits/f2e55492d57bceb23508dff36b70885d3d4e7235)

Reply:
I don't see any harm in the network client construction being lazy. Leaving it as-is but adding additional initialization steps later sounds reasonable to me